### PR TITLE
FM2-223: Add create, update, and delete methods for ServiceRequest

### PR DESCRIPTION
**Pull request by gitcliff**
LINK:https://issues.openmrs.org/browse/FM2-223

### CODE REVIEW

File: omod/src/test/resources/org/openmrs/module/fhir2/providers/ServiceRequestWebTest_UpdateWithWrongId.json [hunk 1]
```diff
@@ -0,0 +1,8 @@
+{
+  "resourceType": "ServiceRequest",
+  "id": "c0258498-2387-77df-97a5-7038c432aaba"
+
+
```
Comment (corneliouzbett) [inline omod/src/test/resources/org/openmrs/module/fhir2/providers/ServiceRequestWebTest_UpdateWithWrongId.json:5]: Remove these extra spaces

File: omod/src/test/resources/org/openmrs/module/fhir2/providers/ServiceRequestWebTest_create.json [hunk 2]
```diff
@@ -0,0 +1,7 @@
+{
+  "resourceType": "ServiceRequest"
+
+
+}
+
```
Comment (corneliouzbett) [inline omod/src/test/resources/org/openmrs/module/fhir2/providers/ServiceRequestWebTest_create.json:6]: Why do you need this file with only resourceType property? It's not a valid ServiceRequest resource CREATE payload?

**Comment by gitcliff** [comment]
@corneliouzbett  @Akayeshmantha  i dont think its possible to introduce the r3 provider for `ServiceRequest `because its not defined in R3 api but only in R4 and R5

**Comment by varung-31** [comment]
> @corneliouzbett @Akayeshmantha i dont think its possible to introduce the r3 provider for `ServiceRequest `because its not defined in R3 api but only in R4 and R5

You can use ProcedureRequest for R3

File: omod/src/test/java/org/openmrs/module/fhir2/providers/r4/ServiceRequestFhirResourceProviderWebTest.java [hunk 3]
```diff
@@ -68,4 +86,93 @@ public void getEncounterByWrongUuid_shouldReturn404() throws Exception {
...
+		MockHttpServletResponse response = put("/ServiceRequest/" + SERVICE_REQUEST_UUID).jsonContent(jsonServiceRequest)
+				.accept(FhirMediaTypes.JSON).go();
+
+		assertThat(response, isBadRequest());
+		assertThat(response.getContentAsString(),
+				containsStringIgnoringCase("body must contain an ID element for update"));
+	}
+
+	@Test
+	public void updateServiceRequest_shouldErrorForIdMissMatch() throws Exception {
```
Comment (varung-31) [inline omod/src/test/java/org/openmrs/module/fhir2/providers/r4/ServiceRequestFhirResourceProviderWebTest.java:146]: Typo in MissMatch

File: omod/src/test/resources/org/openmrs/module/fhir2/providers/ServiceRequestWebTest_update.json [hunk 4]
```diff
@@ -0,0 +1,9 @@
+{
+  "resourceType": "ServiceRequest",
+  "id": "7d13b03b-58c2-43f5-b34d-08750c51aea9"
+
+
```
Comment (varung-31) [inline omod/src/test/resources/org/openmrs/module/fhir2/providers/ServiceRequestWebTest_update.json:5]: Please add more fields to the json files

**Comment by gitcliff** [comment]
@ibacher kindly review

File: omod/src/test/java/org/openmrs/module/fhir2/providers/r4/ServiceRequestFhirResourceProviderWebTest.java [hunk 5]
```diff
@@ -141,4 +160,92 @@ private void verifyUri(String uri) throws Exception {
...
+		MockHttpServletResponse response = put("/ServiceRequest/" + SERVICE_REQUEST_UUID).jsonContent(jsonServiceRequest)
+				.accept(FhirMediaTypes.JSON).go();
+
+		assertThat(response, isBadRequest());
+		assertThat(response.getContentAsString(),
+				containsStringIgnoringCase("body must contain an ID element for update"));
+	}
+
+	@Test
+	public void updateServiceRequest_shouldrRaiseExceptionForIdMissMatch() throws Exception {
```
Comment (varung-31) [inline omod/src/test/java/org/openmrs/module/fhir2/providers/r4/ServiceRequestFhirResourceProviderWebTest.java:219]:
```suggestion
	public void updateServiceRequest_shouldRaiseExceptionForIdMisMatch() throws Exception {
```

File: api/src/main/java/org/openmrs/module/fhir2/providers/r4/ServiceRequestFhirResourceProvider.java [hunk 6]
```diff
@@ -55,6 +63,33 @@ public ServiceRequest getServiceRequestById(@IdParam @NotNull IdType id) {
...
+
+		serviceRequest.setId(id.getIdPart());
+
+		return FhirProviderUtils.buildUpdate(serviceRequestService.update(id.getIdPart(), serviceRequest));
+	}
+
+	@Delete
+	@SuppressWarnings("unused")
+	public OperationOutcome deleteServiceRequest(@IdParam @NotNull IdType id) {
+		org.hl7.fhir.r4.model.ServiceRequest serviceRequest = serviceRequestService.delete(id.getIdPart());
```
Comment (varung-31) [inline api/src/main/java/org/openmrs/module/fhir2/providers/r4/ServiceRequestFhirResourceProvider.java:86]:
```suggestion
		ServiceRequest serviceRequest = serviceRequestService.delete(id.getIdPart());
```

File: api/src/test/java/org/openmrs/module/fhir2/providers/r4/ServiceRequestFhirResourceProviderTest.java [hunk 7]
```diff
@@ -106,6 +112,7 @@ public void getServiceRequestByWithWrongId_shouldThrowResourceNotFoundException(
 	}
 	
 	@Test
+
```
Comment (varung-31) [inline api/src/test/java/org/openmrs/module/fhir2/providers/r4/ServiceRequestFhirResourceProviderTest.java:115]: Please remove this.

File: omod/src/test/java/org/openmrs/module/fhir2/providers/r4/ServiceRequestFhirResourceProviderWebTest.java [hunk 8]
```diff
@@ -141,4 +160,92 @@ private void verifyUri(String uri) throws Exception {
...
+
+	@Test
+	public void createServiceRequest_shouldCreateServiceRequest() throws Exception {
+		String jsonServiceRequest;
+		try (InputStream is = this.getClass().getClassLoader().getResourceAsStream(JSON_CREATE_SERVICE_REQUEST_PATH)) {
+			Objects.requireNonNull(is);
+			jsonServiceRequest = IOUtils.toString(is, StandardCharsets.UTF_8);
+		}
+
+		org.hl7.fhir.r4.model.ServiceRequest serviceRequest = new org.hl7.fhir.r4.model.ServiceRequest();
```
Comment (varung-31) [inline omod/src/test/java/org/openmrs/module/fhir2/providers/r4/ServiceRequestFhirResourceProviderWebTest.java:172]:
```suggestion
		ServiceRequest serviceRequest = new ServiceRequest();
```

File: omod/src/test/java/org/openmrs/module/fhir2/providers/r4/ServiceRequestFhirResourceProviderWebTest.java [hunk 9]
```diff
@@ -141,4 +160,92 @@ private void verifyUri(String uri) throws Exception {
...
+		serviceRequest.setId(SERVICE_REQUEST_UUID);
+
+		when(service.create(any(org.hl7.fhir.r4.model.ServiceRequest.class))).thenReturn(serviceRequest);
```
Comment (varung-31) [inline omod/src/test/java/org/openmrs/module/fhir2/providers/r4/ServiceRequestFhirResourceProviderWebTest.java:175]:
```suggestion
		when(service.create(any(ServiceRequest.class))).thenReturn(serviceRequest);
```

File: omod/src/test/java/org/openmrs/module/fhir2/providers/r4/ServiceRequestFhirResourceProviderWebTest.java [hunk 10]
```diff
@@ -141,4 +160,92 @@ private void verifyUri(String uri) throws Exception {
...
+
+	@Test
+	public void updateServiceRequest_shouldUpdateExistingServiceRequest() throws Exception {
+		String jsonServiceRequest;
+		try (InputStream is = this.getClass().getClassLoader().getResourceAsStream(JSON_UPDATE_SERVICE_REQUEST_PATH )) {
+			Objects.requireNonNull(is);
+			jsonServiceRequest = IOUtils.toString(is, StandardCharsets.UTF_8);
+		}
+
+		org.hl7.fhir.r4.model.ServiceRequest serviceRequest = new org.hl7.fhir.r4.model.ServiceRequest();
```
Comment (varung-31) [inline omod/src/test/java/org/openmrs/module/fhir2/providers/r4/ServiceRequestFhirResourceProviderWebTest.java:191]:
```suggestion
		ServiceRequest serviceRequest = new ServiceRequest();
```

File: omod/src/test/java/org/openmrs/module/fhir2/providers/r4/ServiceRequestFhirResourceProviderWebTest.java [hunk 11]
```diff
@@ -141,4 +160,92 @@ private void verifyUri(String uri) throws Exception {
...
+		serviceRequest.setId(SERVICE_REQUEST_UUID);
+
+		when(service.update(anyString(), any(org.hl7.fhir.r4.model.ServiceRequest.class))).thenReturn(serviceRequest);
```
Comment (varung-31) [inline omod/src/test/java/org/openmrs/module/fhir2/providers/r4/ServiceRequestFhirResourceProviderWebTest.java:194]:
```suggestion
		when(service.update(anyString(), any(ServiceRequest.class))).thenReturn(serviceRequest);
```

File: omod/src/test/java/org/openmrs/module/fhir2/providers/r4/ServiceRequestFhirResourceProviderWebTest.java [hunk 12]
```diff
@@ -141,4 +160,92 @@ private void verifyUri(String uri) throws Exception {
...
+				containsStringIgnoringCase("body must contain an ID element which matches the request URL"));
+	}
+
+	@Test
+	public void deleteServiceRequest_shouldDeleteServiceRequest() throws Exception {
+		OperationOutcome retVal = new OperationOutcome();
+		retVal.setId(SERVICE_REQUEST_UUID);
+		retVal.getText().setDivAsString("Deleted successfully");
+
+		org.hl7.fhir.r4.model.ServiceRequest serviceRequest = new org.hl7.fhir.r4.model.ServiceRequest();
```
Comment (varung-31) [inline omod/src/test/java/org/openmrs/module/fhir2/providers/r4/ServiceRequestFhirResourceProviderWebTest.java:241]:
```suggestion
		ServiceRequest serviceRequest = new ServiceRequest();
```

**Comment by gitcliff** [comment]
@varung-31 i thought the procedure request should be a implemented in a separate pr issue 
cc @ibacher

**Comment by ibacher** [comment]
@gitcliff I'm fine with adding `ProcedureRequest` in a new ticket, but could you address @varung-31's other comments? Thanks for your work on this!

**Comment by gitcliff** [comment]
@ibacher please review ,,thnx

**Pull request merged**